sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateTTLSuite.scala

```diff
@@ -0,0 +1,308 @@
+    key: String,
+    userKey: String,
+    value: Int,
+    isTTLValue: Boolean,
+    ttlValue: Long)
+
+class MapStateTTLProcessor(ttlConfig: TTLConfig)
+  extends StatefulProcessor[String, MapInputEvent, MapOutputEvent]
+    with Logging {
+
+  @transient private var _mapState: MapStateImplWithTTL[String, Int] = _
+
+  override def init(
+      outputMode: OutputMode,
+      timeMode: TimeMode): Unit = {
+    _mapState = getHandle
+      .getMapState("mapState", Encoders.STRING, Encoders.scalaInt, ttlConfig)
+      .asInstanceOf[MapStateImplWithTTL[String, Int]]
+  }
+
+  override def handleInputRows(
+      key: String,
+      inputRows: Iterator[MapInputEvent],
+      timerValues: TimerValues,
+      expiredTimerInfo: ExpiredTimerInfo): Iterator[MapOutputEvent] = {
+    var results = List[MapOutputEvent]()
+
+    for (row <- inputRows) {
+      val resultIter = processRow(row, _mapState)
+      resultIter.foreach { r =>
+        results = r :: results
+      }
+    }
+
+    results.iterator
+  }
+
+  def processRow(
+      row: MapInputEvent,
+      mapState: MapStateImplWithTTL[String, Int]): Iterator[MapOutputEvent] = {
+    var results = List[MapOutputEvent]()
+    val key = row.key
+    val userKey = row.userKey
+    if (row.action == "get") {
+      if (mapState.containsKey(userKey)) {
+        results = MapOutputEvent(key, userKey, mapState.getValue(userKey),
+          isTTLValue = false, -1) :: results
+      }
+    } else if (row.action == "get_without_enforcing_ttl") {
+      val currState = mapState.getWithoutEnforcingTTL(userKey)
+      if (currState.isDefined) {
+        results = MapOutputEvent(key, userKey, currState.get, isTTLValue = false, -1) :: results
+      }
+    } else if (row.action == "get_ttl_value_from_state") {
+      val ttlValue = mapState.getTTLValue(userKey)
+      if (ttlValue.isDefined) {
+        val value = ttlValue.get._1
+        val ttlExpiration = ttlValue.get._2
+        results = MapOutputEvent(key, userKey, value, isTTLValue = true, ttlExpiration) :: results
+      }
+    } else if (row.action == "put") {
+      mapState.updateValue(userKey, row.value)
+    } else if (row.action == "get_values_in_ttl_state") {
+      val ttlValues = mapState.getKeyValuesInTTLState()
+      ttlValues.foreach { elem =>
+        results = MapOutputEvent(key, elem._1, -1, isTTLValue = true, ttlValue = elem._2) :: results
+      }
+    } else if (row.action == "iterator") {
+      val iter = mapState.iterator()
+      iter.foreach { elem =>
+        results = MapOutputEvent(key, elem._1, elem._2, isTTLValue = false, -1) :: results
+      }
+    }
+
+    results.iterator
+  }
+}
+
+class TransformWithMapStateTTLSuite extends TransformWithStateTTLTest {
+
+  import testImplicits._
+  override def getProcessor(ttlConfig: TTLConfig):
+  StatefulProcessor[String, InputEvent, OutputEvent] = {
+    new MapStateSingleKeyTTLProcessor(ttlConfig)
+  }
+
+  def getStateTTLMetricName: String = "numMapStateWithTTLVars"
```

Comment:
Where is this used ? do we have a test verifying the emitted metrics ?

Reply:
Yeah, in the test base, we have a test that makes sure the stateVariable metric is incremented.

Reply:
Could we add `override def` then ?

Reply:
Done